### PR TITLE
MINOR: make LogCleaner.shouldRetainRecord more clear to read

### PR DESCRIPTION
**Pull request by FreeeeLy**
The comments of LogCleaner.shouldRetainRecord are pretty clear to tell us when to retain a record.
However, I think the code is not so clear as the comments so that i make some changes:
`val redundant = foundOffset >= 0 && record.offset < foundOffset`
--->
`val latest= foundOffset < 0 || record.offset >= foundOffset`

if there dose not exist a message with the key, foundOffset will be -1, record.offset is larger than -1:
`val latest = record.offset() >= foundOffset`

### CODE REVIEW

**Comment by FreeeeLy** [comment]
Call for review @cadonna @bbejeck @mjsax @vvcephei @guozhangwang

**Comment by FreeeeLy** [comment]
Retest this,please

**Comment by FreeeeLy** [comment]
JDK11 has one test failed: 
kafka.api.UserQuotaTest.testThrottledProducerConsumer
`java.lang.AssertionError: Client with id=QuotasTestProducer-1 should have been throttled`

JDK 8 passed

**Comment by mjsax** [comment]
Call for review anyof @hachikuji @gwenshap @bob-barrett @cmccabe

File: core/src/main/scala/kafka/log/LogCleaner.scala [hunk 1]
```diff
@@ -735,9 +735,9 @@ private[log] class Cleaner(val id: Int,
        *   1) if there exists a message with the same key but higher offset
        *   2) if the message is a delete "tombstone" marker and enough time has passed
        */
-      val redundant = foundOffset >= 0 && record.offset < foundOffset
+      val latest = record.offset() >= foundOffset
```
Comment (bob-barrett) [inline core/src/main/scala/kafka/log/LogCleaner.scala]: I like this simplification but I think `latest` is still a confusing name since `map.latestOffset` is a separate thing. Maybe name it something like `latestOffsetForKey`?

Comment (FreeeeLy) [reply core/src/main/scala/kafka/log/LogCleaner.scala]: @bob-barrett  done

**Review by bob-barrett — APPROVED** [review]
LGTM, thanks for the patch

File: core/src/main/scala/kafka/log/LogCleaner.scala [hunk 2]
```diff
@@ -735,9 +735,9 @@ private[log] class Cleaner(val id: Int,
        *   1) if there exists a message with the same key but higher offset
        *   2) if the message is a delete "tombstone" marker and enough time has passed
        */
-      val redundant = foundOffset >= 0 && record.offset < foundOffset
+      val latestOffsetForKey = record.offset() >= foundOffset
       val obsoleteDelete = !retainDeletes && !record.hasValue
```
Comment (hachikuji) [inline core/src/main/scala/kafka/log/LogCleaner.scala]: I find this a bit hard to follow also. Is it any clearer if we invert it as well?
```scala
val latestOffsetForKey = record.offset() >= foundOffset
val isRetainedValue = record.hasValue || retainDeletes
latestOffsetForKey && isRetainedValue 
```
Unfortunately I couldn't think of a great name for it.

Comment (FreeeeLy) [reply core/src/main/scala/kafka/log/LogCleaner.scala]: This change makes sense to me because I think logical not operator is sometimes difficult to follow.

Comment (FreeeeLy) [reply core/src/main/scala/kafka/log/LogCleaner.scala]: I push a new commit to invert obsoleteDelete  and change comments to make them conform to the code

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks for the patch!
